In [1]:
import pandas as pd #importing pandas library as pd, like creating an instance of pandas
import matplotlib.pyplot as plt
%matplotlib inline

In [2]:
df_boxes = pd.read_csv('../Common Datasets/boxes.csv') #importing box details datasets
df_problem1 = pd.read_csv('Datasets/problem 1.csv') #importing predict datasets
df_purchase = pd.read_csv('../Common Datasets/purchase.csv') #importing history datasets

# <font color="red">First checking the boxes datasets, is it clean or not</font>

In [3]:
df_boxes.head() #checking collumn and head of datasets

BOX_ID  QUALITY      DELIVERY_OPTION  MILK  MEAT  UNIT_PRICE
0       1  Premium  Home Delivery - CoD   0.0   2.7        9.96
1       2  Premium  Home Delivery - CoD   0.0   2.3       11.96
2       3  Premium  Home Delivery - CoD   0.0   2.4       11.96
3       4  Premium  Home Delivery - CoD   0.0   2.5       11.96
4       5  Premium  Home Delivery - CoD   0.0   2.6       11.96

In [4]:
df_boxes.info() #checking some information

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 290 entries, 0 to 289
Data columns (total 6 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   BOX_ID           290 non-null    int64  
 1   QUALITY          290 non-null    object 
 2   DELIVERY_OPTION  290 non-null    object 
 3   MILK             290 non-null    float64
 4   MEAT             290 non-null    float64
 5   UNIT_PRICE       290 non-null    float64
dtypes: float64(3), int64(1), object(2)
memory usage: 13.7+ KB


=> no null value found and 2 collumn are in object type

In [5]:
df_boxes.describe() #checking min, max, mean, quartile

BOX_ID        MILK        MEAT  UNIT_PRICE
count  290.000000  290.000000  290.000000  290.000000
mean   145.500000    7.567586    2.248966   15.783655
std     83.860002    6.397381    1.151447    4.098064
min      1.000000    0.000000    0.000000    5.560000
25%     73.250000    0.000000    1.800000   12.825000
50%    145.500000    9.000000    2.300000   15.960000
75%    217.750000   12.000000    2.900000   19.980000
max    290.000000   24.000000    5.800000   23.980000

In [6]:
df_boxes['BOX_ID'].nunique()

290

=> so there are 1-290 box in total, no outlier here. full clean

# <font color="red">Checking Problem one dataset</font>

In [7]:
df_problem1.head()

MAGIC_KEY
0  28D5BB06356
1  293BEAB4E98
2  2962EE8065C
3  2957BE29EA9
4  28E351A0745

In [8]:
df_problem1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 58689 entries, 0 to 58688
Data columns (total 1 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   MAGIC_KEY  58689 non-null  object
dtypes: object(1)
memory usage: 458.6+ KB


=> so total 58689 MAGIC_KEY, we need to predict for them

# <font color="red">Checking Purchase history</font>

In [9]:
df_purchase.head()

PURCHASE_DATE    MAGIC_KEY  BOX_ID  BOX_COUNT
0      1/2/2019  2CED678A247    12.0        1.0
1      1/2/2019  2BF58D91BA1    12.0        1.0
2      1/2/2019  2C15B86534E    99.0        1.0
3      1/2/2019  2C32D9A859A     6.0        1.0
4      1/2/2019  2C7A55404D1     4.0        1.0

In [10]:
df_purchase.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2455864 entries, 0 to 2455863
Data columns (total 4 columns):
 #   Column         Dtype  
---  ------         -----  
 0   PURCHASE_DATE  object 
 1   MAGIC_KEY      object 
 2   BOX_ID         float64
 3   BOX_COUNT      float64
dtypes: float64(2), object(2)
memory usage: 74.9+ MB


# <font color="red">Checking Null Value</font>

In [11]:
df_purchase.isnull().sum()

PURCHASE_DATE     0
MAGIC_KEY         0
BOX_ID           47
BOX_COUNT        47
dtype: int64

=> we have some null value, we need to fix these\
=> as we have enough large datasets, we can drop these\
=> in problem 3, we have a magic key that have null value here, so we need to fill this

In [12]:
len(df_purchase)

2455864

In [13]:
box_count_mean = df_purchase['BOX_COUNT'].mean()
box_count_mean = int(box_count_mean)
box_count_mean

1

In [14]:
df_purchase['BOX_COUNT'] = df_purchase['BOX_COUNT'].fillna(box_count_mean)

In [15]:
BOX_ID_Mode = df_purchase['BOX_ID'].mode()[0]
BOX_ID_Mode

163.0

In [16]:
df_purchase['BOX_ID'] = df_purchase['BOX_ID'].fillna(BOX_ID_Mode)

In [17]:
df_purchase.isnull().sum()

PURCHASE_DATE    0
MAGIC_KEY        0
BOX_ID           0
BOX_COUNT        0
dtype: int64

# <font color="red">Checking duplicate</font>

In [18]:
duplicates =  df_purchase[df_purchase.duplicated(keep='first')] #it will return the duplicated row from the datasets
len(duplicates)

77

In [19]:
duplicates

PURCHASE_DATE    MAGIC_KEY  BOX_ID  BOX_COUNT
95145        5/2/2019  2912705D209   230.0        1.0
96588        5/2/2019  2C3ECE53EE6    90.0        1.0
412565      21/2/2019  2979B501C0E   230.0        1.0
440383      23/2/2019  2CD752AE1DD    56.0        1.0
459925      24/2/2019  2BC3E73EA17    56.0        1.0
...               ...          ...     ...        ...
1630632    15/11/2018  2912705D209   268.0        1.0
2129168    11/10/2018  2C4FDB5E96A    89.0        1.0
2202356    18/10/2018  2C3ECE53EE6    89.0        1.0
2243167    17/10/2018  2BC7E532A3A    89.0        1.0
2281942    15/10/2018  2912705D209   233.0        1.0

[77 rows x 4 columns]

In [20]:
df_purchase.loc[(df_purchase.MAGIC_KEY == '2C3ECE53EE6')] #to show the information of male only

PURCHASE_DATE    MAGIC_KEY  BOX_ID  BOX_COUNT
89217        5/2/2019  2C3ECE53EE6    90.0        1.0
96588        5/2/2019  2C3ECE53EE6    90.0        1.0
96589        5/2/2019  2C3ECE53EE6     9.0        1.0
96590        5/2/2019  2C3ECE53EE6   178.0        2.0
2202355    18/10/2018  2C3ECE53EE6    89.0        1.0
2202356    18/10/2018  2C3ECE53EE6    89.0        1.0
2202357    18/10/2018  2C3ECE53EE6   177.0        1.0
2202363    18/10/2018  2C3ECE53EE6     8.0        1.0
2291101    15/10/2018  2C3ECE53EE6   191.0        1.0

In [20]:
df_purchase = df_purchase.drop_duplicates() #this method will return the dataset with no duplicate row
len(df_purchase)

2455787

# <font color="red">Checking outlier</font>

In [21]:
df_purchase.describe()

BOX_ID     BOX_COUNT
count  2.455787e+06  2.455787e+06
mean   2.308835e+02  1.001112e+00
std    2.923347e+04  4.450408e-02
min    1.000000e+00 -1.000000e+00
25%    1.060000e+02  1.000000e+00
50%    1.430000e+02  1.000000e+00
75%    2.150000e+02  1.000000e+00
max    1.111111e+07  1.900000e+01

=> we have some outlier, need to fix these

In [22]:
for column in df_purchase.columns:
    unique_values_count = df_purchase[column].value_counts()
    print(f"Column: {column}")
    print(unique_values_count)
    print()

Column: PURCHASE_DATE
PURCHASE_DATE
28/2/2019     46924
27/2/2019     22311
8/2/2019      21540
24/2/2019     21448
26/2/2019     21409
              ...  
10/11/2018    11475
3/11/2018     10945
20/10/2018    10620
24/11/2018    10454
17/11/2018     9922
Name: count, Length: 151, dtype: int64

Column: MAGIC_KEY
MAGIC_KEY
2CC2011C2DF    40
28D251217BD    38
2BFCB915234    36
29D9005EBAD    35
2A046A740EB    34
               ..
2BF0AECE1BE     1
2CD9631E40C     1
2C3A858F5F9     1
2C9CC96DD72     1
2C4792A5B84     1
Name: count, Length: 1274108, dtype: int64

Column: BOX_ID
BOX_ID
163.0    226595
143.0    189189
162.0    169705
258.0    116822
255.0    100048
          ...  
151.0         1
226.0         1
242.0         1
243.0         1
109.0         1
Name: count, Length: 291, dtype: int64

Column: BOX_COUNT
BOX_COUNT
 1.0     2453348
 2.0        2272
 3.0          83
 4.0          28
-1.0          17
 6.0          17
 5.0          10
 7.0           3
 9.0           2
 19.0          

=> we can see, in BOX_COUNT, there is a value -1, its not possible in real life, its an outlier\
=> 19 is not outlier as other close value still exist. there is no big gap

In [23]:
unique_values_count = df_purchase['BOX_ID'].value_counts()
print(f"Column: {'BOX_ID'}")
print(unique_values_count)
print()

Column: BOX_ID
BOX_ID
163.0    226595
143.0    189189
162.0    169705
258.0    116822
255.0    100048
          ...  
151.0         1
226.0         1
242.0         1
243.0         1
109.0         1
Name: count, Length: 291, dtype: int64



In [24]:
import numpy as np
from scipy import stats

In [25]:
z_scores = np.abs(stats.zscore(df_purchase['BOX_ID']))
threshold = 3
outliers = np.where(z_scores > threshold)[0]
print("Outliers detected using Z-score method:", outliers)

Outliers detected using Z-score method: [    191     192     193     194     195     196     197  525707  597548
  665794  688028 1251185 1834596 1943843 2002835 2105236 2123355]


In [26]:
outliers

array([    191,     192,     193,     194,     195,     196,     197,
        525707,  597548,  665794,  688028, 1251185, 1834596, 1943843,
       2002835, 2105236, 2123355], dtype=int64)

In [27]:
len(outliers)

17

In [28]:
df_purchase.iloc[outliers] #same as df.head(), inner third bracet using as a numpy array

PURCHASE_DATE    MAGIC_KEY      BOX_ID  BOX_COUNT
191          1/2/2019  2CCB122460D  11111111.0       -1.0
192          1/2/2019  2CEDA71E3E9  11111111.0       -1.0
193          1/2/2019  2CBDC970258  11111111.0       -1.0
194          1/2/2019  2932761FEB6  11111111.0       -1.0
195          1/2/2019  28F09714410  11111111.0       -1.0
196          1/2/2019  2BDAE37A00E  11111111.0       -1.0
197          1/2/2019  2C3C7E168F7  11111111.0       -1.0
525713      27/2/2019  2C3C7E168F7  11111111.0       -1.0
597615       1/1/2019  2CEDA71E3E9  11111111.0       -1.0
665861       5/1/2019  2C3C7E168F7  11111111.0       -1.0
688095       6/1/2019  2CBDC970258  11111111.0       -1.0
1251257    10/12/2018  2C3C7E168F7  11111111.0       -1.0
1834669    10/11/2018  2C3C7E168F7  11111111.0       -1.0
1943916    24/11/2018  2CBDC970258  11111111.0       -1.0
2002908     5/10/2018  2CBDC970258  11111111.0       -1.0
2105309    13/10/2018  2BDAE37A00E  11111111.0       -1.0
2123428    12/10/2018  2C3C7E168F7  11111111.0       -1.0

=> both BOX_ID and BOX_COUNT value are inappropriate, we can drop these row

In [29]:
df_purchase = df_purchase.drop(df_purchase.index[outliers])
df_purchase

In [30]:
df_purchase.describe()

BOX_ID     BOX_COUNT
count  2.455770e+06  2.455770e+06
mean   1.539687e+02  1.001126e+00
std    7.043995e+01  4.419170e-02
min    1.000000e+00  1.000000e+00
25%    1.060000e+02  1.000000e+00
50%    1.430000e+02  1.000000e+00
75%    2.150000e+02  1.000000e+00
max    2.900000e+02  1.900000e+01

=> now there is no outlier, MAX for BOX_ID and MIN for BOX_COUNT Fixed

# <font color="red">Now we need to merge the dataset for model</font>

In [31]:
merged_df = pd.merge(df_purchase, df_boxes, on='BOX_ID')

In [32]:
merged_df

PURCHASE_DATE    MAGIC_KEY  BOX_ID  BOX_COUNT  QUALITY  \
0            1/2/2019  2CED678A247    12.0        1.0  Premium   
1            1/2/2019  2BF58D91BA1    12.0        1.0  Premium   
2            1/2/2019  2C15B86534E    99.0        1.0  Premium   
3            1/2/2019  2C32D9A859A     6.0        1.0  Premium   
4            1/2/2019  2C7A55404D1     4.0        1.0  Premium   
...               ...          ...     ...        ...      ...   
2455765    28/10/2018  2BD992B5538    12.0        1.0  Premium   
2455766    28/10/2018  2C97CD72233    17.0        1.0  Premium   
2455767    28/10/2018  2C91C61D372    40.0        1.0  Premium   
2455768    28/10/2018  2CD70CFC4E3    51.0        1.0  Premium   
2455769    28/10/2018  2C4792A5B84     4.0        1.0  Premium   

                        DELIVERY_OPTION  MILK  MEAT  UNIT_PRICE  
0                   Home Delivery - CoD   8.0   1.5       12.98  
1                   Home Delivery - CoD   8.0   1.5       12.98  
2        Delivery from Collection Point   0.0   3.3       13.96  
3                   Home Delivery - CoD   0.0   2.7       11.96  
4                   Home Delivery - CoD   0.0   2.5       11.96  
...                                 ...   ...   ...         ...  
2455765             Home Delivery - CoD   8.0   1.5       12.98  
2455766             Home Delivery - CoD  10.0   1.8       12.98  
2455767             Home Delivery - CoD  12.0   1.8       19.98  
2455768             Home Delivery - CoD  18.0   2.9       23.98  
2455769             Home Delivery - CoD   0.0   2.5       11.96  

[2455770 rows x 9 columns]

In [33]:
merged_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2455770 entries, 0 to 2455769
Data columns (total 9 columns):
 #   Column           Dtype  
---  ------           -----  
 0   PURCHASE_DATE    object 
 1   MAGIC_KEY        object 
 2   BOX_ID           float64
 3   BOX_COUNT        float64
 4   QUALITY          object 
 5   DELIVERY_OPTION  object 
 6   MILK             float64
 7   MEAT             float64
 8   UNIT_PRICE       float64
dtypes: float64(5), object(4)
memory usage: 168.6+ MB


In [34]:
merged_df.describe()

BOX_ID     BOX_COUNT          MILK          MEAT    UNIT_PRICE
count  2.455770e+06  2.455770e+06  2.455770e+06  2.455770e+06  2.455770e+06
mean   1.539687e+02  1.001126e+00  8.132562e+00  2.055554e+00  1.762173e+01
std    7.043995e+01  4.419170e-02  5.777920e+00  9.765884e-01  4.068327e+00
min    1.000000e+00  1.000000e+00  0.000000e+00  0.000000e+00  5.560000e+00
25%    1.060000e+02  1.000000e+00  0.000000e+00  1.800000e+00  1.396000e+01
50%    1.430000e+02  1.000000e+00  1.000000e+01  2.200000e+00  1.798000e+01
75%    2.150000e+02  1.000000e+00  1.100000e+01  2.500000e+00  1.998000e+01
max    2.900000e+02  1.900000e+01  2.400000e+01  5.800000e+00  2.398000e+01

In [35]:
merged_df['PURCHASE_DATE'] = pd.to_datetime(merged_df['PURCHASE_DATE'], format="%d/%m/%Y")

In [36]:
merged_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2455770 entries, 0 to 2455769
Data columns (total 9 columns):
 #   Column           Dtype         
---  ------           -----         
 0   PURCHASE_DATE    datetime64[ns]
 1   MAGIC_KEY        object        
 2   BOX_ID           float64       
 3   BOX_COUNT        float64       
 4   QUALITY          object        
 5   DELIVERY_OPTION  object        
 6   MILK             float64       
 7   MEAT             float64       
 8   UNIT_PRICE       float64       
dtypes: datetime64[ns](1), float64(5), object(3)
memory usage: 168.6+ MB


In [37]:
merged_df.describe()

PURCHASE_DATE        BOX_ID     BOX_COUNT  \
count                        2455770  2.455770e+06  2.455770e+06   
mean   2018-12-20 16:11:50.650264576  1.539687e+02  1.001126e+00   
min              2018-10-01 00:00:00  1.000000e+00  1.000000e+00   
25%              2018-11-11 00:00:00  1.060000e+02  1.000000e+00   
50%              2018-12-24 00:00:00  1.430000e+02  1.000000e+00   
75%              2019-01-30 00:00:00  2.150000e+02  1.000000e+00   
max              2019-02-28 00:00:00  2.900000e+02  1.900000e+01   
std                              NaN  7.043995e+01  4.419170e-02   

               MILK          MEAT    UNIT_PRICE  
count  2.455770e+06  2.455770e+06  2.455770e+06  
mean   8.132562e+00  2.055554e+00  1.762173e+01  
min    0.000000e+00  0.000000e+00  5.560000e+00  
25%    0.000000e+00  1.800000e+00  1.396000e+01  
50%    1.000000e+01  2.200000e+00  1.798000e+01  
75%    1.100000e+01  2.500000e+00  1.998000e+01  
max    2.400000e+01  5.800000e+00  2.398000e+01  
std    5.777920e+00  9.765884e-01  4.068327e+00

In [38]:
merged_df.to_csv('merged_df.csv', index=False)